azurelinuxagent/common/cgroupstelemetry.py

```diff
@@ -13,47 +13,184 @@
 # limitations under the License.
 #
 # Requires Python 2.6+ and Openssl 1.0+
+import threading
+from datetime import datetime as dt
 
-import time
-
-from azurelinuxagent.common import logger
-from azurelinuxagent.common.cgroup import CpuCgroup, MemoryCGroup
 
 class CGroupsTelemetry(object):
     """
     """
-    _tracked = {}
+    _tracked = []
+    _cgroup_metrics = {}
+    _rlock = threading.RLock()
+
+    @staticmethod
+    def _get_metrics_list(metric):
+        return [metric.average(), metric.min(), metric.max(), metric.median(), metric.count(), metric.first_polltime(),
+                metric.last_polltime()]
+
+    @staticmethod
+    def _process_cgroup_metric(cgroup_metrics):
+        current_memory_usage, max_memory_levels, current_cpu_usage = cgroup_metrics.get_metrics()
+        processed_extension = {
+            "memory": {
+                "cur_mem": CGroupsTelemetry._get_metrics_list(current_memory_usage),
+                "max_mem": CGroupsTelemetry._get_metrics_list(max_memory_levels)
+            },
+            "cpu": {
+                "cur_cpu": CGroupsTelemetry._get_metrics_list(current_cpu_usage)
+            }
+        }
+        return processed_extension
 
     @staticmethod
     def metrics_hierarchies():
-        return CGroupsTelemetry._hierarchies
+        pass  # return CGroupsTelemetry._hierarchies
 
     @staticmethod
     def track_cgroup(cgroup):
         """
-        Adds the given item to the list of tracked cgroups
+        Adds the given item to the dictionary of tracked cgroups
         """
-        pass  # TODO
+        # Making track thread-safe.
+        CGroupsTelemetry._rlock.acquire()
+        CGroupsTelemetry._tracked.append(cgroup)
+        # if not CGroupsTelemetry.is_tracked(cgroup.name, cgroup.controller):
+        #     CGroupsTelemetry._tracked[cgroup.name] = {cgroup.controller: cgroup}
+
+        CGroupsTelemetry._rlock.release()
 
     @staticmethod
-    def is_tracked(name):
+    def is_tracked(name, controller):
         """
         Returns true if the given item is in the list of tracked items
         """
-        pass  # TODO
+        return name in CGroupsTelemetry._tracked and controller in CGroupsTelemetry._tracked[name]
 
     @staticmethod
     def stop_tracking(name):
         """
         Stop tracking the cgroups for the given name
         """
-        if CGroupsTelemetry.is_tracked(name):
-            del (CGroupsTelemetry._tracked[name])  # TODO - thread-safeness; review implementation
+        # Pop is atomic and thread-safe.
```

Comment:
> Pop is atomic and thread-safe.

What about append?

If one of them is not thread safe then probably we should protect all of them regardless of thread-safety